### ISSUE TITLE
ListBox keyboard navigation with Virtualizer fails to focus and scroll to last item during async loading

### ISSUE THREAD

**Issue by zzzze** [adobe/react-spectrum#10129]
### Provide a general summary of the issue here

ListBox keyboard navigation does not work correctly with Virtualizer when async loading is enabled.

When using a ListBox with async loading only, pressing the End key moves focus to the last loaded item, scrolls it into view, and triggers loading more items. This can be repeated until all items are loaded.

However, when the ListBox is wrapped with Virtualizer, pressing End several times eventually stops focusing the last item correctly, and the focused item is not scrolled into the visible area.

Reproduction demo:
https://stackblitz.com/edit/d2rrxber?file=src%2FExample.tsx

### 🤔 Expected Behavior?

When ListBox is used with both async loading and Virtualizer, pressing the End key should behave the same as it does without Virtualizer.

Specifically:

Pressing End should move focus to the last currently loaded item.
The focused item should be scrolled into view.
If more items are available, reaching the end should trigger loading more items.
After more items are loaded, pressing End again should move focus to the new last item and scroll it into view.
This should continue to work until there are no more items to load.

### 😯 Current Behavior

When the ListBox is wrapped with Virtualizer, the End key works initially, but after pressing it several times, keyboard navigation breaks.

The last item is no longer focused correctly, and it is not scrolled into the visible area. As a result, users cannot reliably navigate to the end of an async-loaded virtualized list using the keyboard.

### 🖥️ Steps to Reproduce

Open the reproduction demo:
https://stackblitz.com/edit/d2rrxber?file=src%2FExample.tsx

Focus the virtualized ListBox.

Press the End key.

Wait for more items to load.

Press the End key again.

Repeat the previous two steps several times.

Observe that eventually the last item is no longer focused correctly and is not scrolled into view.

### Version

"react-aria-components": "^1.17.0"

### What browsers are you seeing the problem on?

Chrome

### What operating system are you using?

MacOS 26.2

**Comment by LFDanLu** [comment]
I've reproduce this locally as well, seems to be that the `focusedKey` is being properly updated to the last item in the set of newly loaded item, but the key isn't being persisted by the virtualizer. It seems to be that this check https://github.com/adobe/react-spectrum/blob/457647b858c5f271e5f5fad76d9236a34f3515ad/packages/react-stately/src/layout/ListLayout.ts#L281 is returning false and thus we don't update the layout to persist the last of the newest items into the DOM. It seems to be a side effect of https://github.com/adobe/react-spectrum/pull/8533/ since https://github.com/adobe/react-spectrum/blob/457647b858c5f271e5f5fad76d9236a34f3515ad/packages/react-stately/src/virtualizer/OverscanManager.ts#L46-L47 means that we are overscanning in the horizontal direction by 1/3 of the width of the list despite the list only being vertically scrollable, meaning the `requestedRect` is actually `4/3 * width * rect height` rather than being `width * rect height`. This means the `requestedRect` may be prematurely calculated to be larger than the list's full content area (aka the `number of items in the list * item height`) hence 


I think ideally we'd only overscan in the direction that the layout actually scrolls in perhaps, but will need to test that vs just patching the rect comparison check in ListLayout to be something like `this.lastCollection.getItem(key) != null` (aka get the full layout if the persisted key we are looking for is actually in the collection)

**Comment by nwidynski** [comment]
@LFDanLu I think adjusting OverscanManager in such a way will be difficult, since we don't retain ref access in stately. We would have to assert overflow in aria and pass it all the way down - not sure whether that's worth it.

Maybe we can get away with clamping the requested rect by the non-overscanned rect of `this.virtualizer.visibleRect`?

**Comment by LFDanLu** [comment]
@nwidynski yeah, maybe we could have the layouts return what scrollDirection they currently support based on their current orientation and Virtualizer could do that clamping in https://github.com/adobe/react-spectrum/blob/1e5a26596464ae9a5cd9fd60d211f20a955189a1/packages/react-stately/src/virtualizer/Virtualizer.ts#L208-L213 or something?

**Comment by chirokas** [comment]
> maybe we could have the layouts return what scrollDirection they currently support based on their current orientation

For Table, we need to overscan in both directions, but the `orientation` in `TableLayout` is always `'vertical'`, so we can’t determine the scroll direction based on the layout’s current orientation.

<del>
  Instead, we can calculate `scrollDelta` to detect the actual scroll direction
  and determine which axes should be overscanned.
</del>

<del>
```ts
class OverscanManager {
  setVisibleRect(rect) {
    this.scrollDelta.x = rect.x - this.visibleRect.x
    this.scrollDelta.y = rect.y - this.visibleRect.y
  }
  getOverscannedRect() {
    if (this.scrollDelta.y !== 0 || this.scrollDelta.isOrigin()) {
      let overscanY = this.visibleRect.height / 3
      overscanned.height += overscanY
      if (this.velocity.y < 0) {
        overscanned.y -= overscanY
      }
    }
    if (this.scrollDelta.x !== 0 || this.scrollDelta.isOrigin()) {
      let overscanX = this.visibleRect.width / 3
      overscanned.width += overscanX
      if (this.velocity.x < 0) {
        overscanned.x -= overscanX
      }
    }
  }
}
```
</del>

<del>
  I tried this fix last weekend, and it seemed to work well in local testing
  (Firefox).
</del>

**Comment by nwidynski** [comment]
Wouldn't this also mean we may stop overscanning bi-directionally, except at the origin? I understand this is desirable for the specific issue, but fundamentally, OverscanManager's job is to prepare reusables which may enter the view. Architecturally, I feel that OverscanManager is the wrong place to address this issue, since it's too far down the stack.

I think we just need to resolve cross-axis bleed in L281 by comparing the dimensions separately, e.g. `requested.width < content.width || requested.height < content.height`. Removing the check entirely will probably cause infinite loops, as some nodes may still not be laid out after a full pass.

**Comment by chirokas** [comment]
> I think we just need to resolve cross-axis bleed in L281 by comparing the dimensions separately

I tried a similar approach before:

```ts
requestedRect.area < contentSize.area ||
  requestedRect.height < contentSize.height ||
  requestedRect.width < contentSize.width
```

However, I ran into an edge case with `ComboBox` / `Select`: when the menu opens, `menuWidth` hasn’t been updated yet, so `contentSize.width` (`virtualizer.size.width`) is `0` even though `contentSize.height` is already available. At that point, `requestedRect.height` is also `0`.

This could trigger a full layout computation even when the selected item is already in the scroll view.

Maybe something like this would work better:

```ts
requestedRect.area < contentSize.area ||
  (contentSize.area > 0 &&
    (requestedRect.height < contentSize.height ||
      requestedRect.width < contentSize.width))
```

> Architecturally, I feel that OverscanManager is the wrong place to address this issue, since it's too far down the stack.

Just sharing what I ran into - I haven’t thought it through too deeply yet.
